### DISCUSSION TITLE
dark mode for GUI

### DISCUSSION THREAD

**Discussion by Josiah-tan** [opensteno/plover#1415] [General]
I think that it would be a nice feature to have dark mode support for some features like the suggestions, lookup etc.

**Comment by user202729** [comment]
Edit: in newest version refer to the solution below using `plover.qss` file. Just "open config folder", create a new plover.qss file there then paste the content of downloaded Combinear.qss file in.

------

I have a hack like this https://github.com/user202729/plover/commit/0ecc7f05747228dbd5b5170ec998a8136f1f11f4

Not guaranteed to work. Specifically does not work on Windows.

------

Also I think that if qt global style is dark, Plover will also be dark (not sure how to set this)

[edit2: `QT_STYLE_OVERRIDE=Windows plover` or use Fusion will do "something" although not dark... to get all styles do `from PyQt5.QtWidgets import QStyleFactory; QStyleFactory.keys()`]

**Reply by aerickt** [reply]
> I have a hack like this [user202729@0ecc7f0](https://github.com/user202729/plover/commit/0ecc7f05747228dbd5b5170ec998a8136f1f11f4)

Would you happen to know how to apply these tweaks in a qss file? I'm not even sure if there are equivalent selector names and I'm a bit overwhelmed by the [documentation](https://doc.qt.io/qt-6/stylesheet-syntax.html).

This is my attempt so far (slightly different colours):
```
QWidget {
    background-color: rgb(45, 45, 45);
    color: white;
    tooltip-base: black;
    tooltip-text: white;
}

QToolTip {
    background-color: black;
    color: white;
}

QPushButton {
    background-color: rgb(45, 45, 45);
    color: white;
}

QPushButton::bright {
    color: red;
}

QWidget::item {
    color: white;
    background-color: rgb(45, 45, 45);
}

QWidget::item:disabled {
    color: rgb(148, 148, 148);
    background-color: rgb(53, 53, 53);
}

QWidget::item:selected {
    color: black;
    background-color: rgb(42, 130, 218);
}
```

But there are a few annoyances like the highlighted text in the add translation tool being completely unreadable (it's white on white). I've tried adding "QHighlightedText" and "QHighlight" but those don't seem to be the correct selector names.

**Reply by user202729** [reply]
Not sure.

(by the way the full reference of style sheet is in https://doc.qt.io/qt-6/stylesheet-reference.html#list-of-properties , and other links in https://doc.qt.io/qt-6/stylesheet.html -- for qt6 but probably mostly applicable here)

Plover source code has the following...

```
        self._special_fmt = (
            '<span style="' +
            'background-color:' + self.palette().base().color().name() +';' +
            'font-family:monospace;' +
            '">%s</span>'
        )
```

Meanwhile qt documentation...

> Warning: Do not use this function in conjunction with [Qt Style Sheets](https://doc.qt.io/qt-6/stylesheet.html). When using style sheets, the palette of a widget can be customized using the "color", "background-color", "selection-color", "selection-background-color" and "alternate-background-color".

(https://doc.qt.io/qt-6/qwidget.html#palette-prop)

so it might not be possible...? (removing that line works. Didn't check what that line is used for)

(the other annoyance is QCheckBox indicator turns black when the parent item is selected -- Combinear does not have this issue; but then it does not set the `color` of `QWidget::item:selected` either)

**Reply by aerickt** [reply]
Thank you for the insight. I will keep looking into it.

**Comment by Josiah-tan** [comment]
Alright, this is not entirely a solution (just an alternative), but I've created a vim [plugin](https://github.com/Josiah-tan/vim-plover-clippy) that allows you to live view files within vim (meaning that you can have your own custom color scheme)
I specifically built this plugin to view clippy.txt files that are outputted by a plugin [plover clippy](https://github.com/tckmn/plover_clippy) in order to see how I can improve by cutting down unnecessary strokes

**Reply by user202729** [reply]
Related things (disclaimer I didn't try those out)

* [chrisbra/vim-autoread: Use tail -f on a buffer and append new content](https://github.com/chrisbra/vim-autoread/)
* [auto_autoread.vim - Automatically read files when they've changed. Does what 'autoread' promises. : vim online](https://www.vim.org/scripts/script.php?script_id=5206)
* [Can vim monitor realtime changes to a file - Stack Overflow](https://stackoverflow.com/questions/2157914/can-vim-monitor-realtime-changes-to-a-file)

**Comment by antistic** [comment]
In Linux, if you are not using KDE you can follow [this guide in the Arch Linux wiki](https://wiki.archlinux.org/title/Qt#Configuration_of_Qt5_apps_under_environments_other_than_KDE_Plasma) on how to configure the QT theme which Plover should be able to pick up.

**Comment by Josiah-tan** [comment]
As of dev 12 you can use Qt style sheet templates such as this one [here](https://qss-stock.devsecstudio.com/templates.php)
All I had to do was download, unzip and copy Combinear.qss (the style sheet I chose) into my configuration files.
After that, I renamed the file to plover.qss and it worked!
![image](https://user-images.githubusercontent.com/65082231/189676051-9ca0ac34-f68c-4b59-9552-0d36221300ec.png)
